### PR TITLE
[Grouping&Mapping] Properties Refactor

### PR DESCRIPTION
**Pull request by Guillar1**

### CODE REVIEW

File: common/config/rush/pnpm-lock.yaml
Comment (MindaugasButkus) [inline common/config/rush/pnpm-lock.yaml:278]: Don't downgrade the insights-client.

File: packages/itwin/grouping-mapping-widget/package.json
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/package.json:57]: Don't drop the caret.

File: packages/itwin/grouping-mapping-widget/src/widget/components/SelectProject.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/SelectProject.tsx:68]: This change looks unrelated.

File: packages/itwin/grouping-mapping-widget/src/widget/components/HorizontalTile.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/HorizontalTile.tsx:23]: I don't think destructuring helps when you have that many properties. I think simply using `props.onClick` would be good. Now the props do not fit into the screen.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyTable.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyTable.tsx:87]: Is it no longer possible to modify the property? I think you said that you were able to bring this back.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.scss
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.scss:58]: Non-gmw selectors. Where do they come from? What is the background-image below?

File: packages/itwin/grouping-mapping-widget/src/widget/components/SortableHorizontalTile.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/SortableHorizontalTile.tsx:31]: This looks like a hack. What does this do?

File: packages/itwin/grouping-mapping-widget/src/widget/components/SortableHorizontalTile.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/SortableHorizontalTile.tsx:1]: Missing copyright header.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.scss
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.scss:24]: There should be a variable that will match what you need.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:130]: I think it would be better to default to `DataType.String`. Sending `DataType.Undefined` to Reporting API will cause a BadRequest.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:377]: Why the `slice` for Material?

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:344]: Keep the formatting consistent. Here you add each array element to a new line. Below you keep everything on a single line. Pick one.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:143]: What do you mean by this comment? Are these longs the ECInstanceId of the ECClass (in meta.ECClassDef)?

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:145]: In what cases will the `propertyField.properties` array contain more than 1 entry?

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:49]: I get this error: `Module '"@itwin/itwinui-icons-react"' has no exported member 'SvgDragHandleVertical'.`

File: common/config/rush/pnpm-lock.yaml
Comment (Guillar1) [reply common/config/rush/pnpm-lock.yaml]: Forgot to run rush update after merge. Done.

File: packages/itwin/grouping-mapping-widget/src/widget/components/SelectProject.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/SelectProject.tsx:68]: It is unrelated. Small one line bug that stops a search from being made when no input is given.

File: packages/itwin/grouping-mapping-widget/src/widget/components/SortableHorizontalTile.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/SortableHorizontalTile.tsx:31]: Globally change the cursor to grabbing when actively dragging around the tile.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.scss
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.scss]: All of these selectors were for the split library itself. I managed to change it.

File: packages/itwin/grouping-mapping-widget/src/widget/components/SortableHorizontalTile.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/SortableHorizontalTile.tsx:5]: Huh, the git hook didn't run the script to insert it. Manually added it.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.scss
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.scss:25]: Not for vertical spacing, $iui-baseline is typically used. All the other spacing variables are meant to be used in horizontal spacing.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:376]: The property name that Presentation returns is PhysicalMaterial. I slice it out and replace it with Material because that is the actual column name.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx]: Should be fixed.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx]: I believe it happens during property merging. Forgot to consider this case. https://github.com/iTwin/itwinjs-core/blob/master/docs/presentation/content/Terminology.md#property-merging
I wrote the logic to deal with this but I cannot find an iModel where this edge case occurs.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx]: Done.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:350]: Done

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (MindaugasButkus) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx]: This looks like a very weird and unnecessary edge case. See the BisCore schema: https://github.com/iTwin/bis-schemas/blob/master/Domains/0-Core/BisCore.ecschema.xml

PhysicalMaterial everywhere. How can it be the "actual column name"?

File: packages/itwin/grouping-mapping-widget/package.json
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/package.json:110]: Why is split.js a devDependency? I see that react-split has a dependency on split.js, but I don't understand why we need to explicitly add it into our package.json.

Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/package.json:110]: This wasn't suppose to be here. I used it to hoist up the package to check the types when developing this. I'll let you know when this is review ready. I am taking my laptop to another location and I pushed what I had just incase something happens to it.

File: packages/itwin/grouping-mapping-widget/src/widget/components/SelectProject.tsx
Comment (MindaugasButkus) [reply packages/itwin/grouping-mapping-widget/src/widget/components/SelectProject.tsx:68]: Formatting inconsistency - there should be a space between `if` and `(`.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:138]: Rewrite this `for` with a `map`.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:5]: The length of this file is ridiculous. Move out all the property retrieval / construction logic out so that only the UI part is left in this file.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx]: Good catch. I'm not sure how I arrived to "Material". Must have confused another schema when I looked into this a while back.

File: packages/itwin/grouping-mapping-widget/src/widget/components/SelectProject.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/SelectProject.tsx]: done

File: packages/itwin/grouping-mapping-widget/src/widget/components/HorizontalTile.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/HorizontalTile.tsx]: done

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx]: done

File: packages/itwin/grouping-mapping-widget/package.json
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/package.json]: Removed

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx]: Done. And replaced other similar situations with `flatMap`.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx]: Moved out all property constructed relation functions to its own file.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx]: Yes, it is the ECInstanceId of the ECClass stored as a long in memory in the C++ layer of Presentation when dealing with Navigations. Not sure of the intention is. I can link the line this happens if requested due to it being in Azure DevOps.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx]: Should be fixed now really.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:107]: Why is this formatted differently than the line above? They are the same length.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:149]: There is a chance that this might produce unexpected results. I think this would be better:
```
[p.label, p.categoryLabel, p.actualECClassName]
  .map((l) => l.toLowerCase())
  .some((l) => l.includes(activeSearchInput.toLowerCase()))
```

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts:320]: Convert all `propertiesMetaData` with a `map` function once instead of doing the conversion for all of them for each iteration of the outer `while` loop.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:149]: Oh right, forgot to lowercase.

Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:149]: I like this so much better. Thanks :).

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx]: One line again.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts:320]: I will come back to this later as the currently implementation works.

**Comment by Guillar1** [comment]
@MindaugasButkus FYI, I changed where the category was being displayed. It was getting really cramped with long category names so I moved it to the subText of the tile. 
image

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (MindaugasButkus) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx]: Please be a bit more specific what that long is. If it is the value of `ECInstanceId` of the class in the `meta.ECClassDef` table, then say so in the comment.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts:1]: Missing copyright header.

Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts:5]: Done.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts:30]: Instead of adding comments, you could just rename the properties:
```
label -> displayLabel
schema -> sourceSchema
className -> sourceClassName
type -> propertyType
```
etc.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts:173]: Why is there a special case for TypeDefinition?

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts:227]: Is the `schema` and `class` in `fullClassName` always separated by `:`, or can it be also separated by `.`?

File: common/config/rush/pnpm-lock.yaml
Comment (MindaugasButkus) [inline common/config/rush/pnpm-lock.yaml:694]: Why the downgrades?

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:114]: No default provided for the state which cannot be `undefined`. Change the type or provide a default.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:123]: This might need to be a callback, because it depends on `selectedProperties` value.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:274]: Why did you remove the `gmw-` prefix?

File: packages/itwin/grouping-mapping-widget/src/widget/components/HorizontalTile.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/HorizontalTile.tsx:30]: I think it might be better to `!!props.onClickTitle` when setting to the `iui-anchor` property. That way it will get converted to boolean.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts:227]: Always `:`

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts:173]: I have not seen another type of a related instance that weren't the primitive navigations.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:274]: Bad merge brought it back.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx]: Both needed callbacks actually, the `onHandle` props most likely have these functions as dependencies as well.

File: packages/itwin/grouping-mapping-widget/src/widget/components/HorizontalTile.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/HorizontalTile.tsx]: Sure done.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx]: Changed the type. I figured that not passing anything was implicit enough.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts]: Changed them. Let me know what you think.

Comment (MindaugasButkus) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts]: Looks better now.

**Review by MindaugasButkus — APPROVED** [review]

**Pull request merged**
